File: ReactN/src/ShapeAPI.ts
import { v4 as uuidv4 } from "uuid";
import ToolShape from "./ToolShape";

export default class ShapeAPI
{
    static generateShape(input?: ToolShape): ToolShape
    {
        try
        {
            return {
                id: input?.id ?? uuidv4(),
                x: input?.x ?? ShapeAPI.randomCoordinate(7648),
                y: input?.y ?? ShapeAPI.randomCoordinate(4304),
                circle: input?.circle ?? ShapeAPI.randomCircle(),
                opacity: input?.opacity ?? ShapeAPI.randomOpacity(0.0, 1.0),
                rgb: input?.rgb ?? ShapeAPI.randomColor(),
                rotation: input?.rotation ?? ShapeAPI.randomRotation(),
                scale: input?.scale ?? ShapeAPI.randomScale(0.1, 2.0),
                sides: input?.sides ?? ShapeAPI.randomSides(3, 8),
                star: input?.star ?? ShapeAPI.randomStar()
            };
        }
        catch (error)
        {
            console.error("[ShapeAPI] generateShape failed:",error); return {};
        }
    }

    private static randomCircle(): boolean
    {
        try
        {
            return Math.floor(Math.random() * 256) < 25;
        }
        catch (error)
        {
            console.error("[ShapeAPI] randomCircle failed:",error); return false;
        }
    }

    private static randomColor(): string
    {
        try
        {
            const bytes = new Uint8Array(3); crypto.getRandomValues(bytes);

            return `#${bytes[0].toString(16).padStart(2, '0')}${bytes[1].toString(16).padStart(2, '0')}${bytes[2].toString(16).padStart(2, '0')}`;
        }
        catch (error)
        {
            console.error("[ShapeAPI] randomColor failed:",error); return "";
        }
    }

    private static randomCoordinate(max: number): number
    {
        try
        {
            if (max <= 0) { return 0; }

            return Math.random() * max;
        }
        catch (error)
        {
            console.error("[ShapeAPI] randomCoordinate failed:",error); return 0;
        }
    }

    private static randomOpacity(min: number, max: number): number
    {
        try
        {
            return min + (max - min) * Math.random();
        }
        catch (error)
        {
            console.error("[ShapeAPI] randomOpacity failed:",error); return 1.0;
        }
    }

    private static randomRotation(): number
    {
        try
        {
            return Math.random() * 360.0;
        }
        catch (error)
        {
            console.error("[ShapeAPI] randomRotation failed:",error); return 0;
        }
    }

    private static randomScale(min: number, max: number): number
    {
        try
        {
            return min + (max - min) * Math.random();
        }
        catch (error)
        {
            console.error("[ShapeAPI] randomScale failed:",error); return 1.0;
        }
    }

    private static randomSides(min: number, max: number): number
    {
        try
        {
            if (min > max) { return 12; }

            return Math.floor(Math.random() * (max - min + 1)) + min;
        }
        catch (error)
        {
            console.error("[ShapeAPI] randomSides failed:",error); return 12;
        }
    }

    private static randomStar(): boolean
    {
        try
        {
            return Math.floor(Math.random() * 256) < 50;
        }
        catch (error)
        {
            console.error("[ShapeAPI] randomStar failed:",error); return false;
        }
    }
}

File: ReactR/src/shapesvcs/shapeapi.rs
use rand::Rng;
use uuid::Uuid;

use anyhow::{Result, anyhow};
use crate::shapesvcs::toolshape::ToolShape;

pub struct ShapeAPI;

impl ShapeAPI {
    pub fn generate_shape(&self, input: Option<ToolShape>) -> Result<ToolShape> {
        let result = std::panic::catch_unwind(|| {
            let base = input.unwrap_or(ToolShape {
                id: None,
                x: None,
                y: None,
                circle: None,
                opacity: None,
                rgb: None,
                rotation: None,
                scale: None,
                sides: None,
                star: None,
            });
            let shape = ToolShape {
                id: base.id.or(Some(Uuid::new_v4())),
                x: base.x.or(Some(Self::random_coordinate(7648.0))),
                y: base.y.or(Some(Self::random_coordinate(4304.0))),
                circle: base.circle.or(Some(Self::random_circle())),
                opacity: base.opacity.or(Some(Self::random_opacity())),
                rgb: base.rgb.or(Some(Self::random_color())),
                rotation: base.rotation.or(Some(Self::random_rotation())),
                scale: base.scale.or(Some(Self::random_scale(2.0))),
                sides: base.sides.or(Some(Self::random_sides(3, 8))),
                star: base.star.or(Some(Self::random_star())),
            };
            shape
        });
        match result {
            Ok(shape) => Ok(shape),
            Err(_) => Err(anyhow!("generate_shape failed")),
        }
    }

    fn random_circle() -> bool {
        let mut rng = rand::rng();
        rng.random_bool(25.0/256.0)
    }

    fn random_color() -> String {
        let mut rng = rand::rng();
        let r: u8 = rng.random();
        let g: u8 = rng.random();
        let b: u8 = rng.random();
        format!("#{:02X}{:02X}{:02X}", r, g, b)
    }

    fn random_coordinate(max: f64) -> f64 {
        let mut rng = rand::rng();
        rng.random_range(0.0..max)
    }

    fn random_opacity() -> f64 {
        let mut rng = rand::rng();
        rng.random_range(0.0..=1.0)
    }

    fn random_rotation() -> f64 {
        let mut rng = rand::rng();
        rng.random_range(0.0..360.0)
    }

    fn random_scale(max: f64) -> f64 {
        let mut rng = rand::rng();
        rng.random_range(0.1..=max)
    }

    fn random_sides(min: i32, max: i32) -> i32 {
        let mut rng = rand::rng();
        rng.random_range(min..=max)
    }

    fn random_star() -> bool {
        let mut rng = rand::rng();
        rng.random_bool(50.0/256.0)
    }
}